lib/routes/recovery-codes.js

```diff
@@ -0,0 +1,177 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const errors = require('../error')
+const isA = require('joi')
+const validators = require('./validators')
+const HEX_STRING = validators.HEX_STRING
+
+module.exports = (log, db, config, customs, mailer) => {
+  const codeConfig = config.recoveryCodes
+  const RECOVERY_CODE_LENGTH = codeConfig && codeConfig.codeLength || 8
+  const RECOVERY_CODE_COUNT = codeConfig && codeConfig.codeCount || 8
+
+  return [
+    {
+      method: 'GET',
+      path: '/recoveryCodes',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        response: {
+          schema: {
+            recoveryCodes: isA.array().items(isA.string().regex(HEX_STRING))
+          }
+        }
+      },
+      handler(request, reply) {
+        log.begin('replaceRecoveryCodes', request)
+
+        const uid = request.auth.credentials.uid
+        const sessionToken = request.auth.credentials
+        const geoData = request.app.geo
+        const ip = request.app.clientAddress
+        let codes
+
+        customs.check(request, 'replaceRecoveryCodes')
+          .then(replaceRecoveryCodes)
+          .then(sendEmailNotification)
+          .then(emitMetrics)
+          .then(() => reply({recoveryCodes: codes}), reply)
+
+        function replaceRecoveryCodes() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.replaceRecoveryCodes(uid, RECOVERY_CODE_COUNT)
+            .then((result) => {
+              codes = result
+            })
+        }
+
+        function sendEmailNotification() {
+          return db.account(sessionToken.uid)
+            .then((account) => {
+              mailer.sendPostNewRecoveryCodesNotification(account.emails, account, {
+                acceptLanguage: request.app.acceptLanguage,
+                ip: ip,
+                location: geoData.location,
+                timeZone: geoData.timeZone,
+                uaBrowser: request.app.ua.browser,
+                uaBrowserVersion: request.app.ua.browserVersion,
+                uaOS: request.app.ua.os,
+                uaOSVersion: request.app.ua.osVersion,
+                uaDeviceType: request.app.ua.deviceType,
+                uid: sessionToken.uid
+              })
+            })
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'account.recoveryCode.replaced',
+            uid: uid
+          })
+
+          return request.emitMetricsEvent('recoveryCode.replaced', {uid: uid})
+            .then(() => ({}))
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/session/verify/recoveryCode',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            code: isA.string().min(RECOVERY_CODE_LENGTH).max(RECOVERY_CODE_LENGTH).regex(HEX_STRING).required()
```

Comment:
If they're always going to be the same value, we could replace the `.min()` and `.max()` here with a single `.length()` call.